src/types/mappingproxy.rs

```diff
@@ -0,0 +1,614 @@
+    #checkfunction=ffi::PyDictProxy_Check
+);
+
+impl PyMappingProxy {
+    /// Creates a mappingproxy from an object.
+    pub fn new<'py>(
+        py: Python<'py>,
+        elements: &Bound<'py, PyMapping>,
+    ) -> Bound<'py, PyMappingProxy> {
+        unsafe {
+            ffi::PyDictProxy_New(elements.as_ptr())
+                .assume_owned(py)
+                .downcast_into_unchecked()
+        }
+    }
+}
+
+/// Implementation of functionality for [`PyMappingProxy`].
+///
+/// These methods are defined for the `Bound<'py, PyMappingProxy>` smart pointer, so to use method call
+/// syntax these methods are separated into a trait, because stable Rust does not yet support
+/// `arbitrary_self_types`.
+#[doc(alias = "PyMappingProxy")]
+pub trait PyMappingProxyMethods<'py, 'a>: crate::sealed::Sealed {
+    /// Returns a new mappingproxy that contains the same key-value pairs as self.
+    ///
+    /// This is equivalent to the Python expression `self.copy()`.
+    fn copy(&self) -> PyResult<Bound<'_, PyMappingProxy>>;
+
+    /// Checks if the mappingproxy is empty, i.e. `len(self) == 0`.
+    fn is_empty(&self) -> PyResult<bool>;
+
+    /// Returns a sequence containing all keys in the mapping.
+    fn keys(&self) -> PyResult<Bound<'py, PySequence>>;
+
+    /// Returns a sequence containing all values in the mapping.
+    fn values(&self) -> PyResult<Bound<'py, PySequence>>;
+
+    /// Returns a sequence of tuples of all (key, value) pairs in the mapping.
+    fn items(&self) -> PyResult<Bound<'py, PySequence>>;
+
+    /// Returns `self` cast as a `PyMapping`.
+    fn as_mapping(&self) -> &Bound<'py, PyMapping>;
+
+    /// Takes an object and returns an iterator for it. Returns an error if the object is not
+    /// iterable.
+    fn try_iter(&'a self) -> PyResult<BoundMappingProxyIterator<'py, 'a>>;
+}
+
+impl<'py, 'a> PyMappingProxyMethods<'py, 'a> for Bound<'py, PyMappingProxy> {
+    fn copy(&self) -> PyResult<Bound<'_, PyMappingProxy>> {
+        let res = self.call_method0("copy")?;
+        unsafe { Ok(res.downcast_into_unchecked::<PyMappingProxy>()) }
+    }
+
+    fn is_empty(&self) -> PyResult<bool> {
+        Ok(self.len()? == 0)
+    }
+
+    #[inline]
+    fn keys(&self) -> PyResult<Bound<'py, PySequence>> {
+        unsafe {
+            Ok(ffi::PyMapping_Keys(self.as_ptr())
+                .assume_owned_or_err(self.py())?
+                .downcast_into_unchecked())
+        }
+    }
+
+    #[inline]
+    fn values(&self) -> PyResult<Bound<'py, PySequence>> {
+        unsafe {
+            Ok(ffi::PyMapping_Values(self.as_ptr())
+                .assume_owned_or_err(self.py())?
+                .downcast_into_unchecked())
+        }
+    }
+
+    #[inline]
+    fn items(&self) -> PyResult<Bound<'py, PySequence>> {
+        unsafe {
+            Ok(ffi::PyMapping_Items(self.as_ptr())
+                .assume_owned_or_err(self.py())?
+                .downcast_into_unchecked())
+        }
+    }
```

Comment:
It looks like since 3.7 (our minimum supported version) these FFI functions are now guaranteed to return lists: https://docs.python.org/3/c-api/mapping.html#c.PyMapping_Keys

So I think we can use `PyList` as the return type. It would allow for performance gains versus going through the sequence abstraction.

Similarly, would you be willing to update `PyMappingMethods` in the same way (maybe as a separate PR)?

Reply:
Done! Made the changes and opened #4661